### ISSUE TITLE
Allow `process::Child` to be used with external calls to `waitpid` 

### ISSUE THREAD

**Issue by zackw** [rust-lang/rust#39188]
The `wait` and `wait_with_output` methods of `process::Child` expect to do the waiting for the child to exit themselves.  If you are running many processes at once, and you don't know which one will finish first, this is inconvenient, because the system calls `wait` and `waitpid` consume each exit notification as they report it.  (`waitid` can be told not to do that, but it is not fully supported in the libc crate yet.)  That places the `Child` object in an inconsistent state, in which it is not safe to call `wait` or `wait_with_output`.

If `Child` had a `set_exit_status` method, an application in this situation could use it to keep the objects consistent:  something like

```rust
fn wait_all(children: &[&Child]) -> nix::Result<()> {
    use nix::sys::wait::{wait, WaitStatus}
    let mut pending : HashMap<u32, &Child> =
        children.iter().map(|c| (c.id(), c)).collect();
   while pending.len() {
       let status = try!(wait());
       match status {
           Exited(pid, ..) | Signaled(pid, ..) => {
               let child = try!(pending.remove(pid).ok_or(UnexpectedChild(pid)));
               child.set_exit_status(status); 
           },
           _ => unreachable!();
        }
    }
    Ok(())
}
```

(In this hypothetical, `ExitStatus` impls `From<nix::WaitStatus>`, which I believe nix can make happen without assistance from std.)

**Comment by Mark-Simulacrum** [comment]
It's not clear to me that std itself can do much here. Ideally, we'd probably document this somewhere, though I'm not sure where we could do that either. I'm inclined to close because of this, but asking @rust-lang/libs for opinions on the matter.

**Comment by alexcrichton** [comment]
I know I've personally run into this a bunch of times, and @zackw's suggestion of adding a method to libstd seems plausible to me!

**Comment by zackw** [comment]
@Mark-Simulacrum I don't understand why you say "it's not clear to me that std itself can do much"?  I'm only asking for the addition of `Child.set_exit_status`.  Only the stdlib can do that, and that is enough for the stdlib to do.  It is somewhat awkward, admittedly, but it is enough for the likes of tokio to build a clean API around.

**Comment by Mark-Simulacrum** [comment]
I have no idea really what I was saying there... yeah, this seems quite reasonable. I think the best approach here would be to put up a PR implementing this and then we can discuss there and hopefully merge. I don't see much that is controversial here (though not personally hugely familiar with the topic).

**Comment by dtolnay** [comment]
I agree, I think we would be prepared to consider a PR.

**Comment by steveklabnik** [comment]
Triage: i don't believe a PR was ever sent.

**Comment by zackw** [comment]
Yeah, this fell off the bottom of my todo list. I can try to find time in
the next couple weeks to send a PR if there’s still interest.

**Comment by RodBurman** [comment]
Issue [#69843](https://github.com/rust-lang/rust/pull/69843) implies is supersedes and closes this issue, but this issue is not closed after 6 years since the last comment and #69843 itself seems to have been closed with no (or not all) changes implemented. Should this issue be closed now?
